# Вывод первых и последних строк
## Для того, чтобы вывести первые и последние строки, в DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n=5)

In [1]:
#Выведем первые 5 строк таблицы
import pandas as pd
melb_data = pd.read_csv('data\melb_data.csv', sep = ',')
display(melb_data.head())

<>:3: SyntaxWarning: invalid escape sequence '\m'
<>:3: SyntaxWarning: invalid escape sequence '\m'
C:\Users\Кирилл\AppData\Local\Temp\ipykernel_1936\1241064950.py:3: SyntaxWarning: invalid escape sequence '\m'
  melb_data = pd.read_csv('data\melb_data.csv', sep = ',')


index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

# Размерность таблицы
## Чтобы узнать размерность таблицы, можно воспользоваться методом shape, который возвращает кортеж с количеством строк и столбцов

In [2]:
melb_data.shape

(13580, 23)

## Для того, чтобы получить более детальную информацию о столбцах таблицы, можно воспользоваться методом DataFrame info()

In [3]:
melb_data.info()
# Данный метод выводит:
#Информацию об индексах
#Информацию об общем количестве столбцов
#Таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Coint) в каждом столбце
#и типе данных (Dtype), количестве столбцов, в которых используется определенный тип данных
# Количестве оперативной памяти в мегабайтах, которое тратится на хранение данных

# Из вывода info() понятно, что в таблице есть столбец CouncilArea с пропущенными значениями - количество непустых значений меньше, 
# чем количество строк в таблице (12211 < 13580) 

# Пустыми или пропущенными значениями называются значения в ячейках таблицы, которые не заполнены по какой либо причине, то есть
# на их месте стоит пустое место. В Pandas такие значения обозначаются символом NaN (Not-a-Number)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

# Изменение типа данных в столбце
## В некоторых столбцах данные должны отображаться в форме целого числа, но почему-то кодируются числом с плавающей точкой float64, чтобы это исправить можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца

In [4]:
melb_data['Car'] = melb_data['Car'].astype('Int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('Int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('Int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('Int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('Int64')
melb_data.info()
# В данном коде мы при помощи метода astype() последовательно переопределяем столбцы на них же самих, только с другим типом данных int64
# При работе с большим объемом данных данная перекодировка может сыграть большую роль, экономя объем памяти

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  Int64  
 12  Bathroom       13580 non-null  Int64  
 13  Car            13580 non-null  Int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  Int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

# Получение описательной статистики
## Чтобы быстро посмотреть на основные статистические свойства столбцов можно воспользоваться методом describe()

In [5]:
# По умолчанию метод работает с числовыми (int64, float64) столбцами и показывает число непустых значений (count), среднее (mean),
# стандартное отклонение (std), минимальное значение (min), квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%)
# и максимальное значение (max) для каждого столбца исходной таблицы

# Выведем на экран значение статистических параметров только для столбцов Distance, BuildingArea и Price
melb_data.describe().loc[:, ['Distance', 'BuildingArea', 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

In [ ]:
# Выводы:
# Столбец Distance: в наборе данных представлены объекты, удаленные от центра города на расстояние до 48км.
# Столбец BuildingArea: в таблице представлены участки, на которых отсутствует какое-либо здание, (его площадь равна 0), т.е. владельцы
# продавали пустые участки
# Большая часть объектов (75%) имеет площадь здания до 130 кв.м.
# Столбец Price: диапазон цен варьируется от 8.5 до 9 миллионов

## Метод describe() можно применять не только к числовым типам данных, с помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию
Здесь указаны:
-Количество непустых строк (count)
-Количество уникальных значений(unique)
-Самое частое значение(top)
-Частота или объем использования этого значения(freq) для каждого столбца типа object исходной таблицы

In [6]:
melb_data.describe(include=['object'])

Suburb       Address   Type Method SellerG        Date CouncilArea  \
count       13580         13580  13580  13580   13580       13580       12211   
unique        314         13378      3      5     268          58          33   
top     Reservoir  5 Charles St      h      S  Nelson  27/05/2017    Moreland   
freq          359             3   9449   9022    1565         473        1163   

                   Regionname         Coordinates  
count                   13580               13580  
unique                      8               13097  
top     Southern Metropolitan  -37.8361, 144.9966  
freq                     4695                  12

## Выводы:
Столбец Suburb: наибольшее количество проданных объектов (359), находилось в пригороде Reservior
Столбец Type: Самый популярный тип дома - h(коттедж, вилла, дом)
Столбец SellerG(Риелтор): Известно, что всего 268 компаний, но главной является компания Nelson, которая завершила 1565 сделок
Столбец Date: тут хранится информация за последние 58 дней, наибольшее число продаж пришлось на 27.05.2017, 473 сделки 

# Получение частоты уникальных значений в столбце
## Для того, чтобы определить частоту повторения уникального элемента в столбце используется метод value_counts()
Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями - соответствующая им частота

In [7]:
melb_data['Regionname'].value_counts()

Regionname
Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: count, dtype: int64

In [8]:
#Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке этого значения на
#True результат будет представлен в виде ДОЛИ (относительной частоты):
melb_data['Regionname'].value_counts(normalize=True)
# Можно сделать вывод, что наименьшее количество продаж в регионе Western Victoria (менее 1%), а наибольше в Southern Metropolitan

Regionname
Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: proportion, dtype: float64

In [9]:
data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])
print(data.shape)

(3, 2)


In [12]:
melb_data.dtypes

index              int64
Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode           int64
Bedroom            Int64
Bathroom           Int64
Car                Int64
Landsize         float64
BuildingArea     float64
YearBuilt          Int64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount      Int64
Coordinates       object
dtype: object

In [14]:
melb_data.info()
a = 13580 - 12211
print(a)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  Int64  
 12  Bathroom       13580 non-null  Int64  
 13  Car            13580 non-null  Int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  Int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit